### PR TITLE
Rm productinfo

### PR DESCRIPTION
**Pull request by richmcateer**
* Remove DINRegistrar and add DINMarket.
* DIN itself is a product with DIN of 1000000000 (the genesis DIN). Meta.
* Add quantity and info to Order.
* Add orderInfo as a requirement for Markets. This gives a chance for the market to add info (bytes32) to an order which it can then check in isFulfilled to make sure the seller fulfilled the requirement of the order.
* DINMarket converts the registered DIN (uint256) to bytes32 and verifies that the buyer owns this DIN before the buy transaction completes.
* ENSMarket stores the node of the ENS domain as orderInfo and then checks ENS to verify that the buyer is the owner of that node before the buy transaction completes.

### CODE REVIEW

File: contracts/DIN/DINMarket.sol [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+contract DINMarket is PublicMarket, PriceResolver, InventoryResolver, BuyHandler {
+
+	function DINMarket(DINRegistry _dinRegistry, OrderTracker _orderTracker)
+		PublicMarket(_dinRegistry, _orderTracker) {}
+
+	function orderInfo(uint256 DIN) constant returns (bytes32) {
+		uint256 nextDIN = dinRegistry.index() + 1;
+		return bytes32(nextDIN);
+	}
+
```
Comment (richmcateer) [inline contracts/DIN/DINMarket.sol]: This is how DINMarket verifies that a DIN has been registered

File: contracts/DINRegistry.sol [hunk 2]
```diff
@@ -52,32 +46,48 @@ contract DINRegistry {
...
 
     /**
      * Register a new DIN.
-     * @param DIN The DIN to register.
-     * @param owner The account that will own the registered DIN.
+     * @param owner The account that will own the DIN.
      */
-    function registerDIN(uint DIN, address owner) only_registrar only_unregistered(DIN) {
-        records[DIN].owner = owner;
-        NewOwner(DIN, owner, msg.sender);
```
Comment (richmcateer) [inline contracts/DINRegistry.sol:67]: DINRegistry has more control over what DINs get registered. This doesn't allow as much flexibility for registering random numbers in the future, but I'd rather have DINMarket conform to DINRegistry's rules than make its own rules for registering.

File: contracts/ENSMarket/ENSMarket.sol [hunk 3]
```diff
@@ -20,28 +19,37 @@ contract ENSMarket is PublicMarket {
...
+			_orderTracker
+		)
 	{
 		ens = _ens;
 	}
 
+	function orderInfo(uint256 DIN) constant returns (bytes32) {
+		return ENSNode(DIN);
+	}
+
```
Comment (richmcateer) [inline contracts/ENSMarket/ENSMarket.sol:37]: ENSMarket now uses orderInfo to verify purchase fulfillment

File: contracts/OrderTracker.sol [hunk 4]
```diff
@@ -34,7 +36,9 @@ contract OrderTracker {
 		address buyer, 
 		address seller, 
 		uint256 DIN,
-		uint256 value,
```
Comment (richmcateer) [inline contracts/OrderTracker.sol:37]: Add info and quantity to the order

File: contracts/PublicMarket.sol [hunk 5]
```diff
@@ -8,23 +8,23 @@ import './OrderTracker.sol';
...
         InventoryResolver inventoryResolver;    // Returns whether product is in stock.
         BuyHandler buyHandler;                  // Returns the address of the contract that handles orders.
-        bool isValid;                           // True if all above properties are set.
     }
 
     struct Order {
         address buyer;
         address seller;
         uint256 DIN;
-        uint256 amountPaid;
```
Comment (richmcateer) [inline contracts/PublicMarket.sol:27]: Added info and quantity to the order

File: contracts/PublicMarket.sol [hunk 6]
```diff
@@ -62,17 +61,12 @@ contract PublicMarket is Market {
     }
 
     modifier only_correct_price(uint256 DIN, uint256 quantity) {
-        require(totalPrice(DIN, quantity) == msg.value);
```
Comment (richmcateer) [inline contracts/PublicMarket.sol:65]: Different names for Market and PriceResolver

File: contracts/DIN/DINMarket.sol [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+		uint256 lastDIN = firstDIN + quantity - 1;
+
+		// If more than one DIN was registered, check that the first and last DIN are owned by the buyer.
+		return (
+			(dinRegistry.owner(firstDIN) == orders[orderID].buyer) &&
+			(dinRegistry.owner(lastDIN) == orders[orderID].buyer)
+		);
+	}
+
```
Comment (richmcateer) [inline contracts/DIN/DINMarket.sol:37]: This is how DINMarket verifies that a DIN has been registered

**Review by dkindler — APPROVED** [review]
lgtm

**Pull request merged**